node/network/availability-distribution/src/lib.rs

```diff
@@ -218,16 +232,28 @@ impl ProtocolState {
 
 		// register the relation of relay_parent to candidate..
 		for (receipt_hash, fetched) in candidates {
-			let per_candidate = self.per_candidate.entry(receipt_hash).or_default();
-
-			// Cached candidates already have entries and thus don't need this
-			// information to be set.
-			if let FetchedLiveCandidate::Fresh(descriptor) = fetched {
-				per_candidate.validator_index = validator_index.clone();
-				per_candidate.validators = validators.clone();
-				per_candidate.descriptor = descriptor;
-			}
-			per_candidate.live_in.insert(relay_parent);
+			let candidate_entry = match self.per_candidate.entry(receipt_hash) {
+				Entry::Occupied(e) => e.into_mut(),
+				Entry::Vacant(e) => {
+					if let FetchedLiveCandidate::Fresh(descriptor) = fetched {
+						e.insert(PerCandidate {
+							message_vault: HashMap::new(),
+							received_messages: HashMap::new(),
+							sent_messages: HashMap::new(),
+							validators: validators.clone(),
+							validator_index,
+							descriptor,
+							live_in: HashSet::new(),
+							span: jaeger::candidate_hash_span(&receipt_hash, "pending-availability"),
+						})
+					} else {
+						tracing::warn!(target: LOG_TARGET, "No `per_candidate` but not fresh. logic error");
```

Comment:
It seems correct as-is. There's no `per_candidate` since it's a vacant entry. So it would be very strange for `FetchedLiveCandidate` to be `Cached` as there appears to be no bookkeeping for `receipt_hash` yet.

Reply:
Oh yeah. Read it the wrong way round and that more than once :1st_place_medal: